runtime/mount/python_lib/lplots/h5/utils/align.py

```diff
@@ -1,191 +1,191 @@
-    M2 = M.squeeze(0)
-
-    J_img: np.ndarray | None = None
-
-    def normalize_work() -> None:
-        nonlocal J_img
-        J_img = STalign.normalize(np.stack([M2] * 3))         # make RGB-like 3×H_J×W_J
-
-    await capture_output(normalize_work, send, "normalize",
-                         widget_session_key)
-
-    if torch.cuda.is_available():
-        device = "cuda:0"
-    else:
-        device = "cpu"
-
-    # https://nbviewer.org/urls/curioanalysisbioinformatics.s3.us-west-1.amazonaws.com/STalign/STalign_Seeker_mouse_embryo_h%26e.ipynb
-    params = {"L": L,
-              "T": T,
-              "niter": 50,
-              "pointsI": points_I,
-              "pointsJ": points_J,
-              "device": device,
-              "sigmaM": 0.15,
-              "sigmaB": 0.10,
-              "sigmaA": 0.11,
-              "epV": 10,
-              "muB": torch.tensor([0, 0, 0]),
-              "muA": torch.tensor([1, 1, 1])
-    }
-
-    out: dict[str, any] | None = None
-
-    def lddmm_work() -> None:
-        nonlocal out
-        out = STalign.LDDMM(
-            [I_y, I_x], I_img,
-            [J_y, J_x], J_img,
-            **params
-        )
-
-    await capture_output(lddmm_work, send, "lddmm", widget_session_key)
-
-    pts = np.stack([xs, ys], axis=1)  # (N,2) in J-coords
-    dtype = out["A"].dtype
-    pts_torch = torch.tensor(
-        pts,
-        device=device,
-        dtype=dtype
-    )
-
-    tpts = STalign.transform_points_source_to_target(
-        out["xv"], out["v"], out["A"], pts_torch
-    )
-    adata.obsm[new_scatter_data_key] = tpts.cpu().numpy()
+    try:
+        if scatter_data_key not in adata.obsm:
+            await send({
+                "type": "h5",
+                "op": "align_image",
+                "progress": True,
+                "key": widget_session_key,
+                "value": {
+                    "data": {
+                        "stage": "validation",
+                        "error": f"Scatter data key '{scatter_data_key}' not found in adata.obsm"
+                    }
+                }
+            })
+            return
+
+        scatter_data = np.asarray(adata.obsm[scatter_data_key]).tolist()
+
+        subprocess_input = {
+            "scatter_data": scatter_data,
+            "points_i": points_i,
+            "points_j": points_j,
+            "alignment_method": alignment_method,
+            "widget_session_key": widget_session_key
+        }
+
+        if image_bytes is not None:
+            subprocess_input["image_bytes_b64"] = base64.b64encode(image_bytes).decode("utf-8")
+
+        subprocess_script = Path(__file__).parent / "align_subprocess.py"
+
+        # temp file to avoid "Argument list too long" error
+        with tempfile.NamedTemporaryFile(mode="w", suffix=".json", delete=True, encoding="utf-8") as temp_file:
+            json.dump(subprocess_input, temp_file)
+            temp_file.flush()
+            temp_file_path = temp_file.name
```

Comment:
probably want to resolve and pass full path here?

Reply:
the "name" is the path of the tempfile, it is really dumb, but this already is like "/tmp/.....json"